doc/manual/contract.md

```diff
@@ -0,0 +1,175 @@
+
+~~~
+	fun get_at(index: Int): T
+	is
+		ensure(result == self[index])
+	do
+		# Return the element at the given index.
+	end
+~~~
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, using the pop example, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
+
+~~~
+	fun pop(object: T)
+	is
+		ensure(old(size) - 1 == size)
+	do
+		# Remove the last item on the stack
+	end
+~~~
+
+### Additional informations
+
+* When defining a `ensure` or `expect` contract on an attribute, only the automatically generated write property will use the contract.
+
+~~~
+	var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
+~~~
+
+* The `ensure` or `expect` contracts are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+* In `ensure` contracts, the keyword `result` can only be used when the method has a return parameter.
+
+* The expression contained in an `old` is evaluated in the same context as the called method. Thus, it's possible to refer to attributes, methods, and arguments.
+
+* `old(object)` stores a reference to object. By doing so, if `object` is mutated in the method the `old (object)` will also be mutated. To avoid this it's necessary to provide a way to duplicate the object like `clone` method so that `not old (object.clone).is_same_instance (object)` will be true. In the stack example the expresion `old(stack).is_same_instance(stack) && old(stack.size) != old(stack).size` is true. If you need to `old(stack).is_same_instance(stack)` to be false, you need to provide a method to duplicate the object, such as `not old(stack.clone).is_same_instance(stack)` be true.
+
+* Contracts can be declared in several ways, all are equivalent:
+
+~~~
+	fun foo(a: Int, b: Int)
+	is
+		ensure(a > 0, b < 0)
+	do
+		# Do something
+	end
+
+
+	fun foo
+	is
+		ensure(a > 0 and b < 0)
+	do
+		# Do something
+	end
+
+
+	fun foo
+	is
+		ensure(a > 0)
+		ensure(b < 0)
+	do
+		# Do something
+	end
+~~~
+
+
+## Class/interface
+
+### Invariant
+
+Invariants are used to specify the characteristics of a class/interface which must always be true, except when executing a member function. In other words, invariants define conditions that must be respected by our instance. Any instance which does not respect one or more of those conditions will be considered as incoherent. Finally, invariants must hold before and after any method called (except those called by `self` directly).
```

Comment:
How does this work regarding constructors? Are invariants checked right after? Can I have a custom `init`?